src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -0,0 +1,194 @@
+    MetricType& metric,
+    KernelType& kernel) :
+    referenceSet(referenceSet),
+    querySet(querySet),
+    densities(densities),
+    absError(absError),
+    relError(relError),
+    oldFromNewQueries(oldFromNewQueries),
+    metric(metric),
+    kernel(kernel),
+    lastQueryIndex(querySet.n_cols),
+    lastReferenceIndex(referenceSet.n_cols),
+    baseCases(0),
+    scores(0)
+{
+  // Nothing to do.
+}
+
+//! The base case.
+template<typename MetricType, typename KernelType, typename TreeType>
+inline force_inline
+double KDERules<MetricType, KernelType, TreeType>::BaseCase(
+    const size_t queryIndex,
+    const size_t referenceIndex)
+{
+  double distance = metric.Evaluate(querySet.col(queryIndex),
+                                    referenceSet.col(referenceIndex));
+  if (tree::TreeTraits<TreeType>::RearrangesDataset)
+    densities(oldFromNewQueries.at(queryIndex)) += kernel.Evaluate(distance);
+  else
+    densities(queryIndex) += kernel.Evaluate(distance);
+  ++baseCases;
+  lastQueryIndex = queryIndex;
+  lastReferenceIndex = referenceIndex;
+  return distance;
+}
+
+//! Single-tree scoring function.
+template<typename MetricType, typename KernelType, typename TreeType>
+double KDERules<MetricType, KernelType, TreeType>::
+Score(const size_t /* queryIndex */, TreeType& /* referenceNode */)
+{
+  ++scores;
+  traversalInfo.LastScore() = 0.0;
+  return 0.0;
+}
+
+template<typename MetricType, typename KernelType, typename TreeType>
+double KDERules<MetricType, KernelType, TreeType>::Rescore(
+    const size_t /* queryIndex */,
+    TreeType& /* referenceNode */,
+    const double oldScore) const
+{
+  // If it's pruned it continues to be pruned.
+  return oldScore;
+}
+
+//! Double-tree scoring function.
+template<typename MetricType, typename KernelType, typename TreeType>
+inline double KDERules<MetricType, KernelType, TreeType>::
+Score(TreeType& queryNode, TreeType& referenceNode)
+{
+  const double maxKernel =
+    kernel.Evaluate(queryNode.MinDistance(referenceNode));
+  const double minKernel =
+    kernel.Evaluate(queryNode.MaxDistance(referenceNode));
+  const double bound = maxKernel - minKernel;
+  double score;
+
+  if (bound <= (absError + relError * minKernel) / referenceSet.n_cols)
+  {
+    // Auxiliary variables.
+    double kernelValue;
+    arma::vec& referenceCenter = referenceNode.Stat().Centroid();
+    arma::vec& queryCenter = queryNode.Stat().Centroid();
+
+    // If calculating a center is not required.
+    if (tree::TreeTraits<TreeType>::FirstPointIsCentroid)
+    {
+      kernelValue = EvaluateKernel(queryNode.Point(0), referenceNode.Point(0));
+    }
+    // If a child center is the same as its parent center.
+    else if (tree::TreeTraits<TreeType>::HasSelfChildren)
+    {
+      // Reference node.
+      if (referenceNode.Parent() != NULL &&
+          referenceNode.Point(0) == referenceNode.Parent()->Point(0))
+        referenceCenter = referenceNode.Parent()->Stat().Centroid();
```

Comment:
This will actually cause a copy of `referenceCenter` instead of reseating the reference; is that what you intended to happen?  If not I guess a pointer should be used for `queryCenter` and `referenceCenter`.  It's also possible to have multiple self-children, so maybe a `while()` is needed here.  I am not sure it is a problem though, it depends on your intentions when writing it.